RFC: Better inference of `_apply()` (splatting)

martinholters opened the pull request:
Presently, something like
```julia
x::Union{Tuple{Foo,Bar},Tuple{Bork,Baz}}
f(x...)
```
will be inferred as `f(::Any...)`, which is likely suboptimal. This PR tries to improve the situation in two ways:
1 If a `Union` is splatted, inference tries the individual types. In case multiple `Union`s are splatted, this would result in many possible combinations; I just limit to an arbitrarily chosen maximum of 7. Otherwise...
2. a `Union` of several `Tuple`s of the same length is rewritten to a `Tuple` of `Union`s, e.g. `Tuple{Union{Foo, Bork}, Union{Bar, Baz}}` for the example above, if it wasn't handled by 1. already.

As an example, I have asserted the return type of `setindex(::Tuple, ...)` to be a large `Union`. This allows 
```julia
julia> @code_warntype slicedim(ones(1,1), 1, 1)
[...]
  end::Union{Array{Float64,1},Array{Float64,2}}

julia> @code_warntype slicedim(ones(1,1,1,1,1,1), 1, 1)
[...]
  end::Union{Array{Float64,5},Array{Float64,6}}
```
which are the tightest bounds as `slicedim` is actually not type-stable. On master, the first one gives `Union{Array{Float64,1},Array{Float64,2},Float64}`, while for more than 2 dimensions it's just `Any`. (The modification to `setindex` on master would just always give `Any` here).

The code certainly needs some revising, but as I'm completely new to inference, I'd like some feedback on the general idea first. (Blame @vtjnash for encouraging me to even try this!)

**Edit:** This PR has evolved and grown a bit. It now further enhances cases where
* a splatted argument of unknown length does not appear last. A call like `f(1, ones(n::Int)..., "bar")` is now treated like `f(::Int, ::Union{Float64,String}...)`, while on master it is treated like `f(::Any...)`.
* a general iterable is passed. The iteration protocol is simulated to determine the element type produced by the iterable, instead of just always treating it as `Any`.

martinholters committed:
Better inference of `_apply(..., Union{Tuple...}, ...)`

If a Union is splatted, inference tries the individual types, unless
this would exceed MAX_APPLY_UNION_ENUM=8 total tries.  As a stopgap in
that case, a Union of several Tuples of the same length is rewritten to
a Tuple of Unions.

tkelman commented on base/inference.jl:
```diff
@@ -1377,22 +1389,43 @@ function precise_container_types(args, types, vtypes::VarTable, sv)
+    tvnew = Vector{Any}[]
+    for tv in tp, t in tvs[1]
+        push!(tvnew, push!(copy(tv), t))
+        if length(tvnew) > 7
```
should probably make this cutoff a named parameter

martinholters replied:
Definitely. And I have no clue whether 7 is a reasonable choice.

vtjnash commented:
This looks like a great start. I'm not quite sure if `setindex_t` is a good idea, but I think you are definitely on the right track.

martinholters commented:
> This looks like a great start.

Thanks!

> I'm not quite sure if `setindex_t` is a good idea, 

Neither am I, but it was the simplest thing to do to test/showcase the improved `_apply` inference.

>but I think you are definitely on the right track.

I'm just refactoring things a bit. Probably it will look something like having a `precise_container_type` (note the singular) that handles just one of the arguments, and then the loop over the arguments and (if necessary) any `Union` contents inside `abstract_apply`, without the need for the extra `uniontypes_prod`. While doing so, I realized that presently, `precise_container_types` completely bails out if one the arguments is problematic. E.g. with inference having established `x::Int` and `y::Any`, `foo(x, y...)` would be inferred as `foo(::Any...)`. After the refactoring, it should be easy to instead infer this as `foo(:Int, ::Any...)`. Might there be any reason not to do that?

martinholters committed:
Refactor `abstract_apply` and `precise_container_types`

Loop over the arguments in `abstract_apply` and invoke the
modified/renamed `precise_container_type` for each individual argument.
Translate some cases where the container type cannot be inferred
precisely into Vararg's (instead of bailing out and inferring
`f(::Any...)`) and fuse Vararg's with further arguments into a single
trailing Vararg.

martinholters commented:
> Might there be any reason not to do that?

Maybe so. Doing only part of what's possible, all tests still pass locally without significant increase in run time. With one exception, that is:
```
Test (Worker) | Time (s) | GC (s) | GC % | Alloc (MB) | RSS (MB)
subarray (1)  | 4083.07  | 120.84 |  3.0 | 77743.82   | 1621.79 
```
Now this does seem a bit excessive. Let me double-check whether I got something wrong...

martinholters committed:
Simulate iteration protocal when inferring splatted arguments

Closes #20518 and improves inference when splatting genetal iterables.

martinholters commented:
Hm, 32bit travis failed on one the new tests. I wonder whether it's build environment or test/worker association dependent. Build log is backed up [here](https://gist.github.com/martinholters/06698c8841f614089865a0361936cfba), will restart the travis job.

martinholters commented:
Looks like the failure is platform dependent. Time to set up a 32bit VM. Unless someone has a good idea what might be causing this...

martinholters commented:
Amazing that it took me so long to realize that the Travis 32bit build does not use a 32bit (virtual) machine but just cross-compiles...

Anyway, the bug that caused the failure should be fixed (a missing initialization). Also, I added the specialization for `Array` to use `eltype` instead of `abstract_iteration`.

I'm still unsure about the last commit, which adds the `setindex_t` function to add a type-assertion to `setindex(::Tuple, ...)`. While its effect for `slicedim` is quite nice, I wonder whether it's worth it. Should I just remove that commit? Or try whether `setindex` itself can be rewritten to be more inference-friendly without resorting to this monstrous type-assert? Thoughts?

JeffBezanson commented on base/inference.jl:
```diff
@@ -1380,67 +1382,116 @@ function abstract_evals_to_constant(ex, c::ANY, vtypes, sv)
-            end
-        elseif tti <: AbstractArray && i == n
-            result[i] = Any[Vararg{eltype(tti)}]
+# `type` is the inferred type for expression `arg`.
```
`typ` instead of `type`?

martinholters replied:
Yup, will fix.

JeffBezanson commented:
This is some great work!

The huge union in `setindex_t` also makes me a bit nervous. Maybe better to leave it out for now, since it's not needed for the main improvement here.

Seems like this fixes #19957? Could you add a test for that?

martinholters commented:
> Seems like this fixes #19957?

Not completely: it will infer `f(1...)` like `f(::Int...)` instead of `f(::Any)`, but not like `f(::Int)`. Certainly an improvement, but does that count as fixing #19957?

> Could you add a test for that?

To the extent it is actually fixed, yes, will do.

martinholters commented:
Let's see whether this has any effect on the benchmarks:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/20343#issuecomment-279323444) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/9e5bd0b0c6195ae0f5d529fc59ae50666f6b9d5d/0380e24_vs_cee29b0/report.md). cc @jrevels

martinholters commented:
Two of the improvements look real, everything else is noise?

tkelman commented:
the subarray slowdowns look pretty consistent though

KristofferC commented:
Easiest to run again:

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

martinholters commented:
It looks quite consistent indeed, as the same operations seems to be benchmarked twice: https://github.com/JuliaCI/BaseBenchmarks.jl/commit/da9fda347309e3ee47437cfa2e249171109bf79f#commitcomment-20858355

martinholters commented:
Oh boy, `@code_warntype` for the benchmarked `perf_lucompletepivCopy!` gives a 1000+ lines output. But no type-instabilities there. In fact, no `_apply` there. So at first glance, no idea where the slowdown might come.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/20343#issuecomment-279375975) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/fde1f0cd6ed40d5e9d761751ec4856c13d87ab29/0380e24_vs_cee29b0/report.md). cc @jrevels

tkelman commented:
fair enough then
